redwood-protocol-compose/src/commonTest/kotlin/app/cash/redwood/protocol/compose/ProtocolTest.kt

```diff
@@ -80,38 +80,36 @@ class ProtocolTest {
       }
     }
 
-    assertThat(diffClock.awaitDiff()).isEqualTo(
-      Diff(
-        childrenDiffs = listOf(
-          ChildrenDiff.Insert(Id.Root, ChildrenTag.Root, Id(1), WidgetTag(1) /* row */, 0),
-          ChildrenDiff.Insert(Id(1), ChildrenTag(1), Id(2), WidgetTag(3) /* text */, 0),
-          ChildrenDiff.Insert(Id(1), ChildrenTag(1), Id(3), WidgetTag(1) /* row */, 1),
-          ChildrenDiff.Insert(Id(3), ChildrenTag(1), Id(4), WidgetTag(3) /* text */, 0),
-        ),
-        layoutModifiers = listOf(
-          LayoutModifiers(Id(1)),
-          LayoutModifiers(Id(2)),
-          LayoutModifiers(Id(3)),
-          LayoutModifiers(Id(4)),
-        ),
-        propertyDiffs = listOf(
-          PropertyDiff(Id(2), PropertyTag(1) /* text */, JsonPrimitive("hey")),
-          PropertyDiff(Id(4), PropertyTag(1) /* text */, JsonPrimitive("hello")),
-        ),
+    assertThat(changeClock.awaitChanges()).isEqualTo(
+      listOf(
+        Create(Id(1), WidgetTag(1) /* row */),
+        LayoutModifierChange(Id(1)),
+        Create(Id(2), WidgetTag(3) /* text */),
+        LayoutModifierChange(Id(2)),
+        PropertyChange(Id(2), PropertyTag(1) /* text */, JsonPrimitive("hey")),
+        ChildrenChange.Add(Id(1), ChildrenTag(1), Id(2), 0),
+        Create(Id(3), WidgetTag(1) /* row */),
+        LayoutModifierChange(Id(3)),
+        Create(Id(4), WidgetTag(3) /* text */),
+        LayoutModifierChange(Id(4)),
+        PropertyChange(Id(4), PropertyTag(1) /* text */, JsonPrimitive("hello")),
+        ChildrenChange.Add(Id(3), ChildrenTag(1), Id(4), 0),
+        ChildrenChange.Add(Id(1), ChildrenTag(1), Id(3), 1),
+        ChildrenChange.Add(Id.Root, ChildrenTag.Root, Id(1), 0),
```

Comment:
This is probably the best view of both the properties being applied before the widget is added to its parent, but also how the tree is attached from the bottom-up now.